Addon-Vitest: Add channel API to programmatically trigger test runs

JReinhold opened the pull request:
Works on https://github.com/storybookjs/mcp/issues/138

Companion PR: https://github.com/storybookjs/mcp/pull/131

<!-- If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

Added a new programmatic trigger to `addon-vitest`, which allows external actors (like addon-mcp) to trigger a test run and get results back via events on the channel.

To support this, I had to make additional modifications:

1. Add the channel to the `options`, so that any preset property (like `experimental_devServer`) automatically has access to the channel and can do stuff with it.
2. Replaced the flow for how addon-vitest gets the index on the server.
 
Previously, it would _fetch_ the `index.json` when necessary; however, it wouldn't know which origin/url to fetch from. To know that, the manager would send the origin to the server (`indexUrl`) on startup, because the client _does_ know where to fetch the index from. If this sounds bad, it's because it was, but it was a good compromise at the time.

But not anymore, because now you can trigger tests without even opening the manager UI once, which would leave `indexUrl` undefined, and crash the test run. So now it does the more clear thing, which is possible as of recently, but wasn't before. Now `addon-vitest` gets the index directly from `StoryIndexGenerator` on the server, and subscribes to changes to the index, so that when tests are triggered, the index is already available and up-to-date.


<!-- Briefly describe what your PR does -->

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [x] integration tests
- [x] end-to-end tests

#### Manual testing

1. Have a react-vite Storybook with `addon-mcp` installed from https://github.com/storybookjs/mcp/pull/131
2. Start the dev server and configure the MCP server for your agent
3. Prompt your agent with _"Run tests for the Button component stories using the Storybook MCP testing tool. Use the `run-story-tests` tool and report which stories passed or failed."_
4. See that it runs tests and reports results back. See that in the Storybook UI, you can see the statuses too in the sidebar/testing widget, because they are all connected.

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->
This pull request has been released as version `0.0.0-pr-33206-sha-06900f2a`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-33206-sha-06900f2a sandbox` or in an existing project with `npx storybook@0.0.0-pr-33206-sha-06900f2a upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-33206-sha-06900f2a`](https://npmjs.com/package/storybook/v/0.0.0-pr-33206-sha-06900f2a) |
| **Triggered by** | @JReinhold |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`jeppe/addon-vitest-trigger-api`](https://github.com/storybookjs/storybook/tree/jeppe/addon-vitest-trigger-api) |
| **Commit** | [`06900f2a`](https://github.com/storybookjs/storybook/commit/06900f2a11acb34abb46f9cd49642306045e4acd) |
| **Datetime** | Fri Feb  6 12:53:21 UTC 2026 (`1770382401`) |
| **Workflow run** | [21751255283](https://github.com/storybookjs/storybook/actions/runs/21751255283) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/publish.yml) or locally with `gh workflow run --repo storybookjs/storybook publish.yml --field pr=33206`_
</details>
<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Programmatic Vitest test-run API with trigger/response flow, optional config override, and new run/result types
  * A11y addon exposes an opt-in preset and a runtime flag; per-run a11y statuses and reports are surfaced
  * In-memory story index for faster lookup and focused-run filtering

* **Refactor**
  * Standardized accessibility type names
  * Streamlined channel plumbing across server, build, and presets

* **Tests**
  * New tests for selective/focused test triggering and result handling
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

JReinhold committed:
initial triggerTestRun function

JReinhold committed:
add story statuses to trigger api, take screenshots

JReinhold committed:
Merge branch 'next' of github.com:storybookjs/storybook into jeppe/addon-vitest-trigger-api

JReinhold committed:
fix type

JReinhold committed:
try to add statuses to currentRun data structure

JReinhold committed:
Merge branch 'next' of github.com:storybookjs/storybook into jeppe/addon-vitest-trigger-api

JReinhold committed:
migrate to channel based triggering

JReinhold committed:
expose a11yReports on CurrentRun

JReinhold committed:
add Channel to all preset options

JReinhold committed:
update lock file

JReinhold committed:
fix no channel in build

JReinhold committed:
fix missing channel on options everywhere

JReinhold committed:
don't fetch index in vitest, just get it server-side

JReinhold committed:
Merge branch 'next' of github.com:storybookjs/storybook into jeppe/addon-vitest-trigger-api

JReinhold committed:
Merge branch 'next' of github.com:storybookjs/storybook into jeppe/addon-vitest-trigger-api

JReinhold committed:
cleanup, don't crash on getIndex failing

JReinhold committed:
add isAddonA11yEnabled preset property to addon-a11y

JReinhold committed:
support optional config overrides in TRIGGER_TEST_RUN flow

JReinhold committed:
Merge branch 'next' of github.com:storybookjs/storybook into jeppe/addon-vitest-trigger-api

JReinhold committed:
remove unused preview entry from addon-vitest

JReinhold committed:
fix tests and type checks

JReinhold committed:
fix channel type issues

JReinhold commented on code/core/src/core-server/utils/doTelemetry.ts:
```diff
@@ -28,7 +28,11 @@ export async function doTelemetry(
+          ...options,
+          corePresets: [],
+          overridePresets: [],
+        } as unknown as Parameters<typeof sendTelemetryError>[2]['presetOptions'],
```
I'm not a huge fan of these type fixes, that are now necessary becauseo of the new Channel in options. If you have better solutions @kasperpeulen, please let me know.

Here's a summary of the issue:

---

## Root issue

- The failing `core` type checks were caused by a **TypeScript type identity mismatch**, not a runtime bug.
- Two different `Channel` declarations were being mixed in the same typed flows:
  - one from internal/source typing
  - one from generated `dist` typing
- `Channel` has a private field (`sender`), so TypeScript treats them as nominally incompatible even when they look structurally similar.
- That produced errors like:  
  `Type '...Channel' is not assignable to type 'Channel'` and  
  `Types have separate declarations of a private property 'sender'`.

## Why this fix was necessary

- Critical paths (`loadAllPresets`, `Options` objects, telemetry `presetOptions`) require a **single consistent channel type**.
- Without normalization at those boundaries, `nx check core -c production` fails even though runtime behavior is fine.
- The fix aligned typing at boundaries and used explicit boundary casts where TS could not infer safe overlap.
- This was necessary to:
  - remove false-positive type failures,
  - preserve existing runtime behavior,
  - keep CI (`core:check:production`) green.

## What changed (high level)

- Normalized `channel` typing at `loadAllPresets` / `Options` boundaries in core server build/load flows.
- Normalized telemetry `presetOptions` typing in the error-reporting/telemetry call sites.
- No intended behavior changes at runtime; this is a type-system consistency fix.

## Outcome

- `npx nx check core -c production --outputStyle=stream` now passes.
- The fix resolves the private-field `Channel` incompatibility errors that were blocking CI/type-check.

JReinhold committed:
improve test filtering logic to support running focused tests accross multiple story files

kasperpeulen commented on code/addons/vitest/src/types.ts:
```diff
@@ -1,7 +1,11 @@
+import type { PreviewAnnotation, Status, StoryId, StoryIndex } from 'storybook/internal/types';
 import type { API_HashEntry } from 'storybook/internal/types';
 
+// import type { A11yReport } from '@storybook/addon-a11y';
```
`unknown` would be safer than `any` here — it would force callers to narrow before accessing properties, which is better for a type workaround that might get forgotten.

kasperpeulen committed:
Replace Channel class with ChannelLike interface in type declarations

The Channel class has a private `sender` field which causes TypeScript
nominal incompatibility between source and dist declarations. This
introduces a ChannelLike structural interface and uses it in Options,
loadAllPresets, and Builder types, eliminating all `as unknown as`
channel casts.

kasperpeulen committed:
Fix prettier formatting for ChannelLike changes

kasperpeulen reviewed (approved):
LGTM! Seems like some good architecture changes, couple of comments, and I fixed the channel TS issues

kasperpeulen committed:
Fix import ordering in index-json.ts

JReinhold committed:
remove usage of non-exported shouldLog

JReinhold committed:
Merge branch 'next' of github.com:storybookjs/storybook into jeppe/addon-vitest-trigger-api

JReinhold committed:
Move websocket token to a shared singleton to avoid circular dependency between: ws token -> server channel -> loading all presets -> core preset creates ws token -> ...

Co-authored-by: Norbert de Langen <norbert@chromatic.com>
